### PR TITLE
[xaprepare] Add initial reinstall support

### PR DESCRIPTION
**Pull request by pjcollins**
Introduces a 'refresh' argument to provide a way to force reinstallation
of certain programs or dependencies. Usage:

    xaprepare.exe -refresh=sdk,ndk

Also fixes some SDK component caching which was broken:
A few of our Android SDK components do not specify a package revision.
This breaks caching and updating for these components. If a local version
of one of these components is already installed we won't attempt to
replace it if it is out of date. As a result, we've likely been sitting
on out of date platform-tools (adb.exe), and emulator versions on many of
our build/test machines. These caching fixes will hopefully help address
some of the emulator performance and adb.exe related crashes we've seen.

### CODE REVIEW

File: build-tools/xaprepare/xaprepare/Main.cs [hunk 1]
```diff
@@ -101,6 +101,7 @@ static async Task<int> Run (string[] args)
 				{"a|enable-all", "Enable preparation of all the supported targets, ABIs etc", v => parsedOptions.EnableAll = true},
 				{"b|bundle-path=", "Full path to the {{DIRECTORY}} where Xamarin.Android bundle can be found (excluding the file name)", v => parsedOptions.XABundlePath = v?.Trim ()},
 				{"copy-bundle-to=", "Full path to the {{DIRECTORY}} where downloaded bundle should be copied (excluding the file name)", v => parsedOptions.XABundleCopyDir = v?.Trim ()},
+				{"refresh=", "[sdk,ndk] Comma separated list of components which should be reinstalled.", v => parsedOptions.RefreshList = ParseRefreshableComponents (v?.Trim ())},
```
Comment (jonathanpeppers) [inline build-tools/xaprepare/xaprepare/Main.cs]: Do we need a way to trigger this when you queue a build on Azure DevOps? Adding a boolean variable at queue time?

Comment (pjcollins) [reply build-tools/xaprepare/xaprepare/Main.cs]: I was planning to use this in the APK test PR to replace this hack - https://github.com/xamarin/xamarin-android/pull/3042/files#diff-9268c11f675853b241d743a346444d86R263.

I think it makes sense to conditionally add this to other build stages though based on the value of a queue time variable, I'll take a look at adding that.

File: build-tools/xaprepare/xaprepare/Main.cs [hunk 2]
```diff
@@ -308,5 +310,36 @@ static bool ParseBoolean (string value)
 
 			throw new InvalidOperationException ($"Unknown boolean value: {value}");
 		}
+
+		static RefreshableComponent ParseRefreshableComponents (string refreshList)
+		{
+			if (String.IsNullOrEmpty (refreshList))
+				return RefreshableComponent.None;
```
Comment (grendello) [inline build-tools/xaprepare/xaprepare/Main.cs]: This check could be split into two - check for `null` and if it's a match, return `None`; if the string is empty, however, return `All`

File: build-tools/xaprepare/xaprepare/ConfigAndData/Configurables.cs [hunk 3]
```diff
@@ -109,10 +109,10 @@ public static partial class Defaults
...
 			public const string LogFilePrefix = "prepare";
 
-            /// <summary>
-            ///   Default logging verbosity for the entire program. <see cref="Prepare.LoggingVerbosity" />
-            /// </summary>
-            public static readonly LoggingVerbosity LoggingVerbosity = LoggingVerbosity.Normal;
+			/// <summary>
+			///   Default logging verbosity for the entire program. <see cref="Prepare.LoggingVerbosity" />
+			/// </summary>
+			public static readonly LoggingVerbosity LoggingVerbosity = LoggingVerbosity.Normal;
```
[the remark this answers is not in the document]
Comment (pjcollins) [reply build-tools/xaprepare/xaprepare/ConfigAndData/Configurables.cs]: It effectively "fixes" indentation (as this section didn't match the others) but I can revert this portion if you would like?

File: build-tools/xaprepare/xaprepare/Main.cs (hunk 2 above)
Comment (pjcollins) [reply build-tools/xaprepare/xaprepare/Main.cs]: Switched this over to use an optional argument as suggested. If no value is provided this function will never be hit, and the default value of `None` will be used. If `-refresh` or `-refresh=` is provided, `all` will be used.

**Pull request merged**